src/compiler/dataStructures.ts

```diff
@@ -0,0 +1,777 @@
+                return false;
+            }
+        }
+        return true;
+    }
+
+    export function clone<T>(object: T): T {
+        const result: any = {};
+        for (const id in object) {
+            result[id] = (<any>object)[id];
+        }
+        return <T>result;
+    }
+
+    export function extend<T1 extends ObjMap<{}>, T2 extends ObjMap<{}>>(first: T1 , second: T2): T1 & T2 {
+        const result: T1 & T2 = <any>{};
+        for (const id in first) {
+            (result as any)[id] = first[id];
+        }
+        for (const id in second) {
+            if (!hasProperty(result, id)) {
+                (result as any)[id] = second[id];
+            }
+        }
+        return result;
+    }
+
+    export function objMapIsEqualTo<T>(map1: ObjMap<T>, map2: ObjMap<T>): boolean {
+        if (!map1 || !map2) {
+            return map1 === map2;
+        }
+        return containsAll(map1, map2) && containsAll(map2, map1);
+    }
+
+    function containsAll<T>(map: ObjMap<T>, other: ObjMap<T>): boolean {
+        for (const key in map) {
+            if (!hasProperty(map, key)) {
+                continue;
+            }
+            if (!hasProperty(other, key) || map[key] !== other[key]) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /** Copy all entries from `source` to `target`, overwriting any already existing. */
+    export function copyObjMap<V>(source: ObjMap<V>, target: ObjMap<V>): void {
+        forEachValueAndKey(source, (value, key) => {
+            target[key] = value;
+        });
+    }
+}
+
+// Array
+namespace ts {
+    /**
+     * Tests whether a value is an array.
+     */
+    export function isArray(value: any): value is any[] {
+        return Array.isArray ? Array.isArray(value) : value instanceof Array;
+    }
+
+    /**
+     * Iterates through 'array' by index and performs the callback on each element of array until the callback
+     * returns a truthy value, then returns that value.
+     * If no such value is found, the callback is applied to each element of array and undefined is returned.
+     */
+    export function forEach<T, U>(array: T[], callback: (element: T, index: number) => U): U {
+        if (array) {
+            for (let i = 0, len = array.length; i < len; i++) {
+                const result = callback(array[i], i);
+                if (result) {
+                    return result;
+                }
+            }
+        }
+        return undefined;
+    }
+
+    export function contains<T>(array: T[], value: T): boolean {
+        if (array) {
+            for (const v of array) {
+                if (v === value) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    export function indexOf<T>(array: T[], value: T): number {
+        if (array) {
+            for (let i = 0, len = array.length; i < len; i++) {
+                if (array[i] === value) {

```

Comment:
This should check if `i in array && array[i] === value` (or equivalent). Otherwise this fails to skip holes in sparse arrays. See the [canonical polyfill.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf)